rfd/00xx-kubernetes-access-cluster-discovery.md

```diff
@@ -0,0 +1,428 @@
+#restrict-access-to-the-instance-profile-assigned-to-the-worker-node
+
+1. https://goteleport.com/docs/kubernetes-access/guides/standalone-teleport/#step-12-generate-a-kubeconfig
+2. https://docs.aws.amazon.com/eks/latest/APIReference/API_ListClusters.html
+3. https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeCluster.html
+4. https://github.com/kubernetes-sigs/aws-iam-authenticator
+5. https://goteleport.com/docs/setup/guides/joining-nodes-aws-iam/
+6. https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html
+7. https://eksctl.io/usage/iam-identity-mappings/
+8. https://github.com/helm/helm/tree/main/pkg
+9. https://aws.github.io/aws-eks-best-practices/security/docs/iam/#restrict-access-to-the-instance-profile-assigned-to-the-worker-node
+10. https://github.com/gravitational/teleport/blob/7d5b73eda3caf13717a647264032ef983f997c39/api/types/types.proto#L468
+11. https://github.com/gravitational/teleport/blob/d3e33465380de070dd3ec8347c9967c1b1257582/api/client/proto/authservice.proto#L1539
+---
+
+## Alternative enroll method: Helm chart
+
+Teleport has a join method, IAM join, that allows Teleport agents and Proxies to join a Teleport cluster without sharing any secrets when they are running in AWS.
+
+The IAM join method is available to any Teleport agent running anywhere with access to IAM credentials, such as an EC2 instance that is part of an EKS cluster. This method allows any resource that fulfills the defined criteria to be able to join automatically into the Teleport cluster.
+
+To use this method, each agent requires access to `sts:GetCallerIdentity` to use the IAM method. If the operator didn't [block the pod access to IMDS][blockpodaccess], this is already true since the pods inherit the node IAM role, otherwise, it can be granted by IAM OICD Provider.
+
+To configure the IAM joining token method, the operator has to define the IAM token spec.
+
+```yaml
+
+kind: token
+version: v2
+metadata:
+  # the token name is not a secret because instances must prove that they are
+  # running in your AWS account to use this token
+  name: kube-iam-token
+  # set a long expiry time, the default for tokens is only 30 minutes
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  # use the minimal set of roles required
+  roles: [Kube,[App]]
+
+  # set the join method allowed for this token
+  join_method: iam
+
+  allow:
+  # specify the AWS account which nodes may join from
+  - aws_account: "111111111111"
+  # multiple allow rules are supported
+  - aws_account: "222222222222"
+  # aws_arn is optional and allows you to restrict the IAM role of joining nodes
+  - aws_account: "333333333333"
+    aws_arn: "arn:aws:sts::333333333333:assumed-role/teleport-node-role/i-*"
+```
+
+Once Teleport has discovered a cluster and granted access to its API, it installs the Helm Agent chart via [Helm library][helmlib]. The Teleport Helm chart has to be updated to support IAM joining token.
```

Comment:
Even with this method we'd run into the problem with initial access grant to Teleport, right? Since the discovery agent needs to have access to be able to install Helm chart in the cluster.

Reply:
Yes, we have exactly the same issue mentioned like in the other method.